controllers/storageoscluster/node-manager_operand.go

```diff
@@ -0,0 +1,337 @@
+		return true, nil
+	}
+
+	log.V(4).Info("node-manager not ready")
+	return false, nil
+}
+
+func (c *NodeManagerOperand) PostReady(ctx context.Context, obj client.Object) error {
+	c.currentStateLock <- true
+	currState := c.currentState
+	<-c.currentStateLock
+
+	// Skip if not deployed.
+	if !currState {
+		return nil
+	}
+
+	ctx, span, _, log := instrumentation.Start(ctx, "NodeManagerOperand.PostReady")
+	defer span.End()
+
+	watcher := watchers.NewDaemonSetWatcher(c.appsGetter.DaemonSets(obj.GetNamespace()), "storageos node")
+	if err := watcher.Setup(ctx, true, "app.kubernetes.io/component=control-plane"); err != nil {
+		log.Error(err, "unable to set daemonset watcher")
+		return err
+	}
+
+	if err := c.updateReplicas(ctx, obj, log); err != nil {
+		return err
+	}
+
+	watcher.Start(ctx, func(watchChan <-chan watch.Event) error {
+		log.Info("watcher has started")
+		for {
+			select {
+			case event, ok := <-watchChan:
+				if !ok {
+					log.Info("watcher has closed")
+					return errors.New("watcher has closed")
+				} else if event.Type != watch.Modified {
+					continue
+				}
+
+				log.Info("daemonset has changed")
+
+				err := func() (err error) {
+					ctx, cancel := context.WithTimeout(ctx, time.Minute)
+					defer cancel()
+
+					ticker := time.NewTicker(time.Second)
+					defer ticker.Stop()
+
+					for {
+						select {
+						case <-ticker.C:
+							if err = c.updateReplicas(ctx, obj, log); err != nil {
+								log.Error(err, "unable to update replicas")
+								continue
+							}
+
+							log.Info("successfully updated replicas")
+							cancel()
+						case <-ctx.Done():
+							return
+						}
+					}
+				}()
+				if err != nil {
+					return err
+				}
+			case <-c.watchCloseChan:
+				log.Info("resource has deleted")
+				return nil
+			}
+		}
+	})
+
+	return nil
+}
+
+func (c *NodeManagerOperand) Ensure(ctx context.Context, obj client.Object, ownerRef metav1.OwnerReference) (eventv1.ReconcilerEvent, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "NodeManagerOperand.Ensure")
+	defer span.End()
+
+	var err error
+	c.getCurrentState.Do(func() {
+		ctx, cancel := context.WithTimeout(ctx, time.Minute)
+		defer cancel()
+
+		// Get current state of the deployment object.
+		nodeManagerDep := &appsv1.Deployment{}
+		key := client.ObjectKey{Name: nmDeploymentName, Namespace: obj.GetNamespace()}
+		err = c.client.Get(ctx, key, nodeManagerDep)
+		c.currentState = err == nil
+
+		log.V(4).Info("node-manager state", "deployed", c.currentState)
+	})
+	if err != nil && !apierrors.IsNotFound(err) {
+		span.RecordError(err)
+		return nil, err
+	}
+
+	b, err := getNodeManagerBuilder(c.fs, obj, c.kubectlClient)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+
+	cluster, ok := obj.(*storageoscomv1.StorageOSCluster)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert %v to StorageOSCluster", obj)
+	}
+
+	if len(cluster.Spec.NodeManagerFeatures) > 0 && !c.currentState {
```

Comment:
Note to myself: Maybe this `!c.currentState` isn't thread safe, depends on the behaviour of operator toolkit.